src/main.cpp

```diff
@@ -44,51 +44,48 @@ class any_error_reporter {
   static any_error_reporter make(output_format format) {
     switch (format) {
     case output_format::gnu_like:
-      return any_error_reporter(text_error_reporter(std::cerr));
+      return any_error_reporter(new text_error_reporter(std::cerr));
     case output_format::vim_qflist_json:
-      return any_error_reporter(vim_qflist_json_error_reporter(std::cout));
+      return any_error_reporter(new vim_qflist_json_error_reporter(std::cout));
     }
     QLJS_UNREACHABLE();
   }
 
   void set_source(padded_string_view input, const file_to_lint &file) {
-    std::visit(
-        [&](auto &r) {
-          using reporter_type = std::decay_t<decltype(r)>;
-          if constexpr (std::is_base_of_v<vim_qflist_json_error_reporter,
-                                          reporter_type>) {
-            r.set_source(input, file.path, file.vim_bufnr);
-          } else {
-            r.set_source(input, file.path);
-          }
-        },
-        this->reporter_);
+    text_error_reporter *text_reporter = dynamic_cast<
+      text_error_reporter *>(this->tape_.get_reporter());
+    vim_qflist_json_error_reporter *vim_reporter = dynamic_cast<
+      vim_qflist_json_error_reporter *>(this->tape_.get_reporter());
```

Comment:
I want to avoid `dynamic_cast` if possible. In this case, I think a separate `bool` or `enum` variable would be better than `dynamic_cast`. Or we could use `std::variant` holding pointers.

Alternatively, we could make `error_tape` a template, and make the underlying `error_reporter` type a template parameter. Then we could keep the old style of using `std::variant`. For example:
```cpp
using reporter_variant =
    std::variant<error_tape<text_error_reporter>, error_tape<vim_qflist_json_error_reporter>>;
```
If you go down the template route, I recommend having the `error_tape` construct the underlying `error_reporter`, and perfect-forwarding the constructor arguments.

Reply:
On my side, I tried to use template, but I had trouble converting `reporter_variant` to `error_reporter *`.
I can understand template but `std::variant` is a bit obscure for me... I will try to implement what you suggested, this should help me converting the variant to the true type.

Reply:
I finally managed to go with a complete template way.
Please let me know if there are still problems (e.g. I had linking errors when template functions were defined in `error-tape.cpp` so I defined them in `error-tape.h`).